Add the plugin install engine

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

As part of [the removal of the Legacy REST API from WooCommerce core in v9.0](https://developer.woo.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/), the separate extension will be automatically installed in the sites in which the API is active.

This pull request introduces a `PluginInstaller` class with a `install_plugin` method that will do the following:

- Use an instance of `Plugin_Upgrader` to install a plugin from the specified URL.
- Add information about the installed plugin (name, version, current date and optional metadata) in a `woocommerce_autoinstalled_plugins` option.
- Hook on `after_plugin_row` to display a "This plugin was installed by WooCommerce" notice in the plugins list page, with an optional "More information" link.

![image](https://github.com/woocommerce/woocommerce/assets/937723/3f50dab4-881a-4b3c-9bb0-f831ebb478dc)

- Hook on `upgrader_process_complete` to remove from the option the information about any plugin that is manually reinstalled or upgraded.

For security the following restrictions are enforced:

- The plugin source must be the WordPress.org plugins directory, that is, the source URL must start with `https://downloads.wordpress.org/`
- If the name of the installer is `WooCommerce` (which is the default if no installer name is provided), the caller of the `install_plugin` method must be code inside the WooCommerce core `includes` or `src` directory.

At this time `PluginInstaller` only supports installing new extensions, it can't be used to reinstall or upgrade already installed extensions. Also it only installs plugins, it does not activate them.

### How to test the changes in this Pull Request:

Before starting, if you have the _WooCommerce Legacy REST API_ plugin installed, uninstall it first.

#### Testing the plugin autoinstall

1. Temporarily add the following to the `wc-core-functions.php` file:

```
function wc_install_plugin($name, $metadata = array()) {
    return wc_get_container()->get(\Automattic\WooCommerce\Internal\Utilities\PluginInstaller::class)->install_plugin($name, $metadata);
}
```

1. Open a shell with `wp shell` and execute the following:

```php
wc_install_plugin('https://downloads.wordpress.org/plugin/woocommerce-legacy-rest-api.1.0.1.zip');
```

The result should look like this:

```
=> array(3) {
  ["messages"]=>
  array(5) {
    [0]=>
    string(121) "Downloading installation package from https://downloads.wordpress.org/plugin/woocommerce-legacy-rest-api.1.0.1.zip&#8230;"
    [1]=>
    string(106) "The authenticity of woocommerce-legacy-rest-api.1.0.1.zip could not be verified as no signature was found."
    [2]=>
    string(28) "Unpacking the package&#8230;"
    [3]=>
    string(28) "Installing the plugin&#8230;"
    [4]=>
    string(30) "Plugin installed successfully."
  }
  ["plugin_name"]=>
  string(59) "woocommerce-legacy-rest-api/woocommerce-legacy-rest-api.php"
  ["install_ok"]=>
  bool(true)
}
```

3. Verify that he option has been created:

```
wp> get_option('woocommerce_autoinstalled_plugins')
=> array(1) {
  ["woocommerce-legacy-rest-api/woocommerce-legacy-rest-api.php"]=>
  array(2) {
    ["version"]=>
    string(5) "1.0.1"
    ["date"]=>
    string(19) "2024-02-21 12:10:13"
  }
}
```

4. Go to WooCommerce - Status - Logs, see that a `plugin_auto_installs` log has been created and it contains an entry like this:

![image](https://github.com/woocommerce/woocommerce/assets/937723/49424842-34d7-4ac9-a5a0-e64a9e67fe36)

5. Go to the plugins list and search the WooCommerce Legacy REST API plugin. Verify that it has actually been installed (but not activated) and a notice appears under its name:

![image](https://github.com/woocommerce/woocommerce/assets/937723/3f50dab4-881a-4b3c-9bb0-f831ebb478dc)

The notice won't appear if you add the following snippet to your WooCommerce install:

```php
add_filter( 'woocommerce_show_autoinstalled_plugin_notices', '__return_false' );
```

#### Testing error conditions

6. Try to install the plugin again with the same shell command, this time you'll get:

```
=> array(2) {
  ["messages"]=>
  array(5) {
    [0]=>
    string(121) "Downloading installation package from https://downloads.wordpress.org/plugin/woocommerce-legacy-rest-api.1.0.1.zip&#8230;"
    [1]=>
    string(28) "Unpacking the package&#8230;"
    [2]=>
    string(28) "Installing the plugin&#8230;"
    [3]=>
    string(110) "Destination folder already exists. /home/konamiman/wordpress/./wp-content/plugins/woocommerce-legacy-rest-api/"
    [4]=>
    string(27) "Plugin installation failed."
  }
  ["install_ok"]=>
  bool(false)
}
```

7. Try to install a plugin specifying an invalid URL but still within the WordPress.org domain (you can e.g. modify the URL adding an "x" at the end), this time you'll get:

```
=> array(2) {
  ["messages"]=>
  array(2) {
    [0]=>
    string(122) "Downloading installation package from https://downloads.wordpress.org/plugin/woocommerce-legacy-rest-api.1.0.1.zipx&#8230;"
    [1]=>
    string(26) "Download failed. Not Found"
  }
  ["install_ok"]=>
  bool(false)
}
```

8. Try to install a plugin specifying an URL that is not in the WordPress.org domain (e.g. just `http://example.com/SomePlugin.zip`), this time you'll get:

```
InvalidArgumentException: Only installs from the WordPress.org plugins directory (plugin URL starting with 'https://downloads.wordpress.org/') are allowed.
```

9. Try to invoke the install method directly, without an intermediate function inside WooCommerce core:

```php
wc_get_container()->get(Automattic\WooCommerce\Internal\Utilities\PluginInstaller::class)->install_plugin('https://downloads.wordpress.org/plugin/woocommerce-legacy-rest-api.1.0.1.zip');
```

You will get:

```
InvalidArgumentException: If the value of 'installed_by' is 'WooCommerce', the caller of the method must be a WooCommerce core class or function.
```

These attempts will also generate log entries, but this time of "error" type:

![image](https://github.com/woocommerce/woocommerce/assets/937723/10fab854-41db-4599-b0bb-734b250ae9d9)

Note however that this only applies to failed attempts that return error information (exceptions will be logged in the usual fatal errors log if they are part of a web request).

#### Testing custom metadata

10. Uninstall the plugin and install it again, this time with this command:

```php
wc_get_container()->get(Automattic\WooCommerce\Internal\Utilities\PluginInstaller::class)->install_plugin('https://downloads.wordpress.org/plugin/woocommerce-legacy-rest-api.1.0.1.zip', ['installed_by' => 'a nerd', 'info_link' => 'https://woo.com/']);
```

This time the direct (not via WooCommerce caller) install will work because of the `installed_by` metadata.

11. Verify that the notice that appears in the plugins list now has a different "installed by" text and has also a "More information" link:

![image](https://github.com/woocommerce/woocommerce/assets/937723/31eea406-6f70-4950-96b0-8741a6034d02)

#### Testing manual reinstall/upgrade

12. With the Legacy REST API plugin having been installed via the `PluginInstaller` class, reinstall it manually from the plugins page using the "Add new plugin" - "Upload plugin" buttons. You can get the zip file for the plugin [here](https://wordpress.org/plugins/woocommerce-legacy-rest-api/).

13. Reload the plugins page and verify that the notice under the plugin name no longer appears.

14. Verify that the option that stores the autoinstalled plugins information has been deleted:

```
$ wp option get woocommerce_autoinstalled_plugins
Error: Could not get 'woocommerce_autoinstalled_plugins' option. Does it exist?
```

If other plugins have been autoinstalled then the option will continue to exist, but the information for the specific plugin that has been manually reinstalled/upgraded will have been removed.

**Note:** The information won't be removed from the option if the plugin is uninstalled. This shouldn't be an issue, though: if the plugin is reinstalled, the option will be updated (or deleted) accordingly.

#### Testing in multisite

15. Create a multisite WordPress install. Create a few secondary sites.

16. Install WooCommerce with this pull request applied, but activate only in some of the sites.

17. Go to the sites where WooCommerce is active and configure the logger to write log entries in the database (in WooCommerce - Status - Logs - Settings). _This is needed to overcome a bug in the current filesystem logger: it will always write logs in the main site even if a secondary site is selected with `switch_to_blog`._

18. Repeat the main test (successfully installing the plugin) and verify that the log entry is written in all the sites in which WooCommerce is active.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Konamiman committed:
Add the PluginInstaller class

Konamiman committed:
Add changelog file

Konamiman committed:
Fix linting issues

jorgeatorres commented on plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php:
```diff
@@ -0,0 +1,317 @@
+		}
+
+		include_once ABSPATH . 'wp-admin/includes/file.php';
+		include_once ABSPATH . 'wp-admin/includes/misc.php';
```
I think it'd be good to document _why_ we need these includes as it's not entirely clear (at least to me). If we're using a function from those files, it'd be good to add a comment indicating that's why we're including them.

Konamiman replied:
On closer inspection turns out that these includes aren't actually needed, so I have removed them.

jorgeatorres commented on plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php:
```diff
@@ -0,0 +1,317 @@
+
+		include_once ABSPATH . 'wp-admin/includes/file.php';
+		include_once ABSPATH . 'wp-admin/includes/misc.php';
+		include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+		include_once ABSPATH . 'wp-admin/includes/class-automatic-upgrader-skin.php';
```
For extra safety I believe we could _detect_ whether the classes we need from these files already exist before attempting the include, for which we should also check that the file actually exists, [similar to what WP-CLI does](https://github.com/wp-cli/wp-cli/blob/4c68c264b3aa5a987291f36882ad71387c28f997/php/utils-wp.php#L156-L166).

Even though we have an `include_once()` it _might_ be the case that those classes change locations in WP at some point in the future. Maybe I'm worrying too much but it wouldn't hurt to cleanly bail out if we can't properly load the files and classes we need for this function.

Konamiman replied:
Ok, I'm now checking if the classes exist before executing the includes.

jorgeatorres commented on plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php:
```diff
@@ -0,0 +1,317 @@
+		$upgrader = new \Plugin_Upgrader( $skin );
+
+		$install_ok = $upgrader->install( $plugin_url );
+		wp_clean_plugins_cache();
```
`$upgrader->install()` already clears the cache after a successful install, so we might not need to do this. See https://developer.wordpress.org/reference/classes/plugin_upgrader/install/.

jorgeatorres commented on plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php:
```diff
@@ -0,0 +1,317 @@
+		wp_clean_plugins_cache();
+		$result = array( 'messages' => $skin->get_upgrade_messages() );
+		if ( $install_ok ) {
+			$new_plugins      = get_plugins();
+			$new_plugin_names = array_keys( $new_plugins );
+			if ( is_null( $plugin_name ) ) {
+				$diff_plugins = array_diff( $new_plugin_names, $old_plugin_names );
+				if ( empty( $diff_plugins ) ) {
+					$install_ok = false;
+				} else {
+					$plugin_name    = current( $diff_plugins );
+					$plugin_version = $new_plugins[ $plugin_name ]['Version'];
+				}
+			} else {
+				$plugin_info = $new_plugins[ $plugin_name ] ?? null;
+				if ( is_null( $plugin_info ) ) {
+					$install_ok = false;
+				} else {
+					$plugin_version = $plugin_info['Version'];
+				}
+			}
```
I might be missing something here, but why can't we trust the result of `$upgrader->install()`? Why do we need to fetch the plugin name again this way / what would be the use case?

If we need to check something, we could maybe rely on `Plugin_Upgrader::plugin_info()` which should gives us the path to the file with plugin information. Similarly, there's some other information available in the (public) instance variable `$result` in `Plugin_Upgrader`.
Not sure if the diff approach is very solid.

Konamiman replied:
No, it's not you who is missing something, it's me: I just hadn't noticed the existence of the `plugin_info` method 🤦🏻‍♂️ So I have replaced all this code with a simple call to this method.

jorgeatorres commented on plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php:
```diff
@@ -0,0 +1,317 @@
+			return;
+		}
+
+		foreach ( get_sites() as $site ) {
+			switch_to_blog( $site->blog_id );
+			try {
+				$callback();
+			} finally {
+				restore_current_blog();
+			}
+		}
```
Not sure this will perform well on big networks. I understand this is done to set the option and also log that the plugin was installed. I'm not entirely sure about the latter, but for the former, couldn't we use `get_site_option()` in the MU context instead of saving an option to each site in the network?

Konamiman replied:
We need to do the blog switch for the logging anyway, and it doesn't look like `get_site_option` would provide a big performance gain compared to `switch`/`restore` as it's doing a lot of processing too: https://developer.wordpress.org/reference/functions/get_network_option/

jorgeatorres reviewed (changes requested):
Hi @Konamiman!

Thanks for your work on this 🥇 . Despite how tricky this is, the API looks good to me overall and tests well both on single and MU setups 💯.

I left some comments on the code, both to tighten checks a bit and also to clarify a few things.

Please let me know what you think.

Konamiman committed:
Changes in the installer:

- Allow installS from the WordPress.org directory only
- If the installer name is WooCommerce, throw if the caller is not
  in 'includes' or 'src' in WooCommerce core
- Remove two unnecessary includes
- Check if WP plugin upgrader classes already exist before
  executing the includes for them
- Get the installed plugin info using the plugin_info method
  of the plugin upgrader class

Konamiman committed:
Remove the get_installed_plugin_info method as it's no longer used

jorgeatorres commented on plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php:
```diff
@@ -0,0 +1,283 @@
+
+		foreach ( get_sites() as $site ) {
+			switch_to_blog( $site->blog_id );
+			try {
```
Should we check that WC is active on those sites before attempting to run an arbitrary callback? I believe this is not a problem with the _current_ callbacks as they will just silently fail at using the WC logger, thanks being wrapped in `try {}`, but this might not necessarily be ok depending on what other possible callbacks do exactly.

I don't feel this is a blocker at this point, but maybe something to consider. What are your thoughts @Konamiman?

Konamiman replied:
> Should we check that WC is active on those sites before attempting to run an arbitrary callback?

That's a good point, I have pushed a commit doing that.

jorgeatorres reviewed (approved):
Hey @Konamiman!

Thanks for updating things as requested. I'm pre-approving as I have tested that things work well both on single site and multisite setups, but I did leave a comment that I'd like us to discuss before this is finally merged.

Thanks!

Konamiman committed:
Log plugin install only on sites with WooCommerce active
